### ISSUE TITLE
Assign a table to a TableCollection

### ISSUE THREAD

**Issue by benjeffery** [tskit-dev/tskit#1489]
Currently this results in an error:
```python
>>> tc = tskit.TableCollection(1)
>>> tc.individuals=tskit.IndividualTable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
```
Is there any reason that it shouldn't be possible? The implementation of this would seem to be clearing the existing table and copying the table data across.

Came up in the context of https://github.com/tskit-dev/tskit/pull/1487#discussion_r648387416

**Comment by petrelharp** [comment]
Makes sense to me!

**Comment by petrelharp** [comment]
However, since all these tables refer to each other, it doesn't seem that common that we'd actually have an IndividualTable that makes sense by itself - although I see the counter-example in #1487.

**Comment by benjeffery** [comment]
I agree, we have a lot of code that does the copy-clear-mutate-add/append dance. I prefer making a new table then assigning it as in your first example.

**Comment by jeromekelleher** [comment]
Sure, it makes sense. We would need to be clear that the semantics is this though:
```
tc.individuals = new_table
assert tc.individuals == new_table
assert tc.individuals is not new_table
```

The fact that we're not actually assigning the object itself is what put me off this in the past, as that seems a bit counter intuitive and contrary to what you might think of the semantics of assignment. The object ownership stuff holding the tables in a TableCollection is very tricky, and I definitely don't want to go messing with that.

**Comment by jeromekelleher** [comment]
I guess another way of doing this would be to have a method ``replace`` or something, which does the same thing, which doesn't require us getting into these murky waters:
```
tc.individuals.replace(new_table)
assert tc.individuals == new_table
assert tc.individuals is not new_table
```

I'm not sure about the name, but "Replace the contents of this table with the contents of the specified table" seems like a clear definition?

**Comment by benjeffery** [comment]
I agree that it would be better to have this be a named method - adding a helpful message to the error on assignment that signposts the named function. How about `replace_with`?

**Comment by jeromekelleher** [comment]
Yes, that sounds good.

**Comment by petrelharp** [comment]
If we don't just get to assign with `=` then I'm not sure there'd be a big advantage over the current way:
```
old_muts = tc.mutations.copy()
tc.mutations.clear()
for mut in old_muts:
  tc.mutations.add_row( ... )
```

**Comment by jeromekelleher** [comment]
No harm in having two ways of doing it, though. It would make it more natural to pass around tables on their own.

**Comment by hyanwong** [comment]
Actually, should it be as simple as this in the BaseTable class:

```
    def replace_with(self, other):
        """
        Entirely replace this table with another
        """
        if type(self) != type(other):
            raise ValueError("replacement column is not of the same type")
        tables.nodes.set_columns(
            metadata_schema = other.metadata_schema,
            **{column: other.getattr(column) for column in self.column_names}
        )
```

**Comment by benjeffery** [comment]
Given that we have `replace_with` and assignment gives a very nice error message I'm closing this.
